src/Microsoft.OData.Core/JsonLight/ODataJsonLightDeltaReader.cs

```diff
@@ -166,36 +171,32 @@ public override bool Read()
         }
 
         /// <summary> Asynchronously reads the next <see cref="Microsoft.OData.ODataItem" /> from the message payload. </summary>
-        /// <returns>A task that when completed indicates whether more items were read.</returns>
-        public override Task<bool> ReadAsync()
+        /// <returns>
+        /// A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains true if more items were read; otherwise false.
+        /// </returns>
+        public override async Task<bool> ReadAsync()
         {
-            return this.underlyingReader.ReadAsync().FollowOnSuccessWith(t =>
+            var result = await this.underlyingReader.ReadAsync()
+                .ConfigureAwait(false);
+
+            if (this.underlyingReader.State == ODataReaderState.DeletedResourceStart)
             {
-                if (this.underlyingReader.State == ODataReaderState.DeletedResourceStart)
+                while ((result = await this.underlyingReader.ReadAsync().ConfigureAwait(false)) && this.underlyingReader.State != ODataReaderState.DeletedResourceEnd)
                 {
-                    this.SkipToDeletedResourceEnd();
+                    // Skip to end of a deleted resource
+                    this.SetNestedLevel();

```

Comment:
Why do we call `this.SetNestedLevel()` here? In the existing version, the `SkipToDeletedResourceEnd()` method does not call `this.SetNestedLevel()`

Reply:
@habbes The previous implementation was unwieldy/buggy starting with the void async method (`private async void SkipToDeletedResourceEnd()`) that made it not awaitable - HIGHLY discouraged. If you look at the revised implementation it mirrors the synchronous implementation and I added tests to verify behaviour unlike before.

Reply:
Here's how the synchronous `Read` is implemented:
```csharp
        public override bool Read()
        {
            // Translate state transitions to legacy states
            bool response = this.underlyingReader.Read();
            if (this.underlyingReader.State == ODataReaderState.DeletedResourceStart)
            {
                while ((response = this.underlyingReader.Read()) && this.underlyingReader.State != ODataReaderState.DeletedResourceEnd)
                {
                    // skip to end of a deleted resource
                    this.SetNestedLevel();
                }
            }
            this.SetNestedLevel();
            return response;
        }
```

Reply:
@gathogojr does that mean the existing `ReadAsync()` implementation has a bug? Since it seems to be different the `Read()`.

Reply:
@habbes That would be my bet. And it would have gone unnoticed due to reason of async API being rarely used. If you look at the test file, the commented tests I got rid of were for the async API

Reply:
> What is the `SkipToDeletedResourceEnd()` method supposed to do?

It was a poorly designed asynchronous method (`void` return type) that would call the `ReadAsync` method of the underlying reader (the `ODataJsonLightReader` in this case) and itself recursively until it gets to the end of the deleted resource.